examples/cfd/swe_distributed_gnn/train.py

```diff
@@ -0,0 +1,580 @@
+        acc_loss = 0
+        model.train()
+
+        for train_batch in tqdm(range(cfg.data.num_examples), leave=False):
+            inp, tar = get_random_data(dataloader.dataset, model.device)
+            with amp.autocast(enabled=cfg.model.enable_amp):
+                prd = model(inp)
+                for _ in range(cfg.data.num_future):
+                    prd = model(prd)
+                loss = l2loss(prd, tar)
+
+            acc_loss += loss.item() * inp.size(0)
+            optimizer.zero_grad(set_to_none=True)
+            if dist_manager.rank == 0:
+                gscaler.scale(loss).backward()
+            gscaler.step(optimizer)
+            gscaler.update()
+
+            iters += 1
+
+        acc_loss = acc_loss / len(dataloader.dataset)
+
+        dataloader.dataset.set_num_examples(cfg.data.num_valid)
+
+        # perform validation
+        valid_loss = 0.0
+        model.eval()
+        with torch.no_grad():
+            for val_batch in range(cfg.data.num_valid):
+                inp, tar = get_random_data(dataloader.dataset, model.device)
+                prd = model(inp)
+                for _ in range(cfg.data.num_future):
+                    prd = model(prd)
+                loss = l2loss(prd, tar)
+                valid_loss += loss.item() * inp.size(0)
+
+        valid_loss = valid_loss / len(dataloader.dataset)
+
+        epoch_time = time.time() - epoch_start
+
+        if metrics is not None:
+            current_lr = optimizer.param_groups[0]["lr"]
+            metrics["epochs"].append(epoch)
+            metrics["current_lr"].append(current_lr)
+            metrics["epoch_time"].append(epoch_time)
+            metrics["acc_loss"].append(acc_loss)
+            metrics["val_loss"].append(valid_loss)
+
+        scheduler.step()
+
+    train_time = time.time() - train_start
+
+    if dist_manager.rank == 0:
+        logger.info(f"Training took {train_time}.")
+        logger.info("Training Metrics")
+        for k, v in metrics.items():
+            logger.info(f"{k}: {v}")
+
+
+@hydra.main(
+    version_base="1.3", config_path=".", config_name="config_graphcast_swe.yaml"
+)
+def main(cfg: DictConfig):
+    # realistic FP32 / TF32 settings
+    torch.backends.cuda.matmul.allow_tf32 = True
+    torch.backends.cudnn.allow_tf32 = True
+    torch.set_float32_matmul_precision("high")
+
+    # default of 0.703125 gives (256, 512)
+    input_res = (
+        int(180.0 / cfg.data.angular_resolution),
+        int(360.0 / cfg.data.angular_resolution),
+    )
+
+    torch.manual_seed(cfg.seed)
+    torch.cuda.manual_seed(cfg.seed)
+
+    # initialize DistributedManager and define process group
+    # across which graph is partitioned and model is parallel across
+    # for now, based
+    DistributedManager.initialize()
+    if DistributedManager().distributed:
+        graph_partition_pg_name = "graph_partition"
+        world_size = torch.distributed.get_world_size()
+        DistributedManager.create_process_subgroup(
```

Comment:
The current example e.g. doesn't use DDP, i.e. only model-parallelism. Technically, we don't even need to create the process group for the graph partition. This is rather just to show that one might want to use a dedicated process group for that.
Your call on whether we want to show this in this simplified fashion or by using `ProcessGroupConfig`.

Reply:
I think it would be good to use `ProcessGroupConfig` to define the graph partitions in the model implementation as well as move the mark as shared into the model (though it's trivial in this case). That way, we can also use this model and example as a template for other users that want to build model parallel architectures. Thoughts?